Source/NSSpeechRecognizer.m

```diff
@@ -23,8 +23,279 @@
+#import <Foundation/NSDistributedNotificationCenter.h>
+#import <Foundation/NSDebug.h>
+#import <Foundation/NSUUID.h>
+#import "GSFastEnumeration.h"
+#import "AppKit/NSWorkspace.h"
 
-@implementation NSSpeechRecognizer
+id   _speechRecognitionServer = nil;
+BOOL _serverLaunchTested = NO;
+
+#define SPEECH_RECOGNITION_SERVER @"GSSpeechRecognitionServer"
+
+@interface NSObject (SpeechRecognitionServerPrivate)
+
+- (void) addToBlockingRecognizers: (NSString *)s;
+- (void) removeFromBlockingRecognizers: (NSString *)s;
+- (BOOL) isBlocking: (NSString *)s;
+- (void) addClient;
 
 @end
 
+@implementation NSSpeechRecognizer
+
++ (void) initialize
+{
+  if (self == [NSSpeechRecognizer class])
+    {
+      // Test for an existant server...
+      _speechRecognitionServer =
+        [NSConnection rootProxyForConnectionWithRegisteredName: SPEECH_RECOGNITION_SERVER
+                                                          host: nil];
+
+      // if none exists, start one.  We will connect with it in init.
+      if (nil == _speechRecognitionServer)
+        {
+          NSWorkspace *ws = [NSWorkspace sharedWorkspace];
+          [ws launchApplication: SPEECH_RECOGNITION_SERVER
+                       showIcon: NO
+                     autolaunch: NO];
+        }
+    }
+}
+
+- (void) _restartServer
+{
+  if (nil == _speechRecognitionServer && !_serverLaunchTested)
+    {
+      unsigned int i = 0;
+      
+      // Wait for up to five seconds  for the server to launch, then give up.
+      for (i = 0 ; i < 50 ; i++)
+        {
+          _speechRecognitionServer = [NSConnection
+                                           rootProxyForConnectionWithRegisteredName: SPEECH_RECOGNITION_SERVER
+                                                                               host: nil];
+          RETAIN(_speechRecognitionServer);
+          if (nil != _speechRecognitionServer)
+            {
+              NSDebugLog(@"Server found!!!");
+              break;
+            }
+          [NSThread sleepForTimeInterval: 0.1];
+        }
+      
+      // Set a flag so we don't bother waiting for the speech recognition server to
+      // launch the next time if it didn't work this time.
+      _serverLaunchTested = YES;
+    }
+  
+  if (_speechRecognitionServer == nil)
+    {
+      NSLog(@"Cannot restart speech recognition server.");
+    }
+}
+
+- (void) processNotification: (NSNotification *)note
+{
+  NSString *word = (NSString *)[note object];
+
+  if (_listensInForegroundOnly)
+    {
+      if (_appInForeground == NO)
+        {
+          return;
+        }
+    }
+  
+  if (_blocksOtherRecognizers)
+    {
+      NS_DURING
+        {
+          if ([_speechRecognitionServer isBlocking: [_uuid UUIDString]] == NO)
+            {
+              // If we are not a blocking recognizer, then we are blocked...
+              return;
+            }
+        }
+      NS_HANDLER
+        {
+          NSLog(@"%@", localException);
+          [self _restartServer];
```

Comment:
This is some error handling, but not sufficient. If the connection is broken you should reconnect and retry the same call. That is why Richard suggested the performXXX methods. That way a retry is fairly simple.

Reply:
At this point the server works.   Overbuilding it for reliability doesn't seem super necessary since it's not a mission critical process.   I will take this as a note for the next release along with the bundles for both the recognizer and the synthesizer.